libnvme/ctrl.c

```diff
@@ -0,0 +1,439 @@
+_psd_getter_func_gen_uint32_t(exlat);
+_psd_getter_func_gen_uint8_t(rrt);
+_psd_getter_func_gen_uint8_t(rrl);
+_psd_getter_func_gen_uint8_t(rwt);
+_psd_getter_func_gen_uint8_t(rwl);
+_psd_getter_func_gen_uint16_t(idlp);
+_psd_getter_func_gen_uint8_t(ips);
+_psd_getter_func_gen_uint16_t(actp);
+_psd_getter_func_gen_uint8_t(apw);
+_psd_getter_func_gen_uint8_t(aps);
+
+static const char *_nvme_ctrl_data_prop_ascii(uint8_t *data, size_t size)
+{
+	char *rc_str = NULL;
+
+	assert(data != NULL);
+	assert(size != 0);
+
+	rc_str = (char *) calloc(sizeof(char), size + 1);
+	if (rc_str == NULL) {
+		errno = ENOMEM;
+		return NULL;
+	}
+
+	/* Remove trailing white space */
+	while(size > 0) {
+		if ((data[size - 1] == ' ') || (data[size - 1] == '\0'))
+			--size;
+		else
+			break;
+	}
+
+	if (size > 0)
+		memcpy(rc_str, data, size);
+
+	errno = 0;
+	return rc_str;
+}
+
+int _nvme_ctrl_new(const char *nvme_ctrl_dev_path, struct nvme_ctrl **cnt,
+		   char *err_msg)
+{
+	int rc = NVME_OK;
+	int fd = -1;
+
+	assert(nvme_ctrl_dev_path != NULL);
+	assert(cnt != NULL);
+
+	*cnt = (struct nvme_ctrl *) calloc(sizeof(struct nvme_ctrl), 1);
+	_alloc_null_check(err_msg, *cnt, rc, out);
+
+	(*cnt)->dev_path = strdup(nvme_ctrl_dev_path);
+
+	_alloc_null_check(err_msg, (*cnt)->dev_path, rc, out);
+
+	_good(_nvme_ioctl_fd_open(nvme_ctrl_dev_path, &fd, err_msg), rc, out);
+
+	_good(_nvme_ioctl_identify(fd, (uint8_t *) &((*cnt)->raw_data),
+				   _NVME_ADMIN_CMD_IDENTIFY_CNS_ALL_CTRL,
+				   0,
+				   err_msg),
+	      rc, out);
+
+out:
+	if (fd >= 0)
+		close(fd);
+	if (rc != NVME_OK) {
+		nvme_ctrl_free(*cnt);
+		*cnt = NULL;
+	}
+	return rc;
+}
+
+
+int nvme_ctrl_get_by_fd(int fd, struct nvme_ctrl **cnt, const char **err_msg)
+{
+	char err_msg_buff[_NVME_ERR_MSG_BUFF_LEN];
+	int rc = NVME_OK;
+	char *tmp_err_msg = NULL;
+
+	assert(cnt != NULL);
+
+	_nvme_err_msg_clear(err_msg_buff);
+	*cnt = NULL;
+	if (err_msg != NULL)
+		*err_msg = NULL;
+
+	*cnt = (struct nvme_ctrl *) calloc(sizeof(struct nvme_ctrl), 1);
+	_alloc_null_check(err_msg_buff, *cnt, rc, out);
+
+	_good(_nvme_ioctl_identify(fd, (uint8_t *) &((*cnt)->raw_data),
+				   _NVME_ADMIN_CMD_IDENTIFY_CNS_ALL_CTRL,
+				   0,
+				   err_msg_buff),
+	      rc, out);
+
+out:
+	if (rc != NVME_OK) {
+		nvme_ctrl_free(*cnt);
+		*cnt = NULL;
+		if (err_msg != NULL) {
+			tmp_err_msg = (char *)
+				malloc(sizeof(char *) *
+				       _NVME_ERR_MSG_BUFF_LEN);
+			if (tmp_err_msg != NULL) {
+				memcpy(tmp_err_msg, err_msg_buff,
+				       _NVME_ERR_MSG_BUFF_LEN);
+				*err_msg = tmp_err_msg;
+			}
+		}
+	}
+	return rc;
+
+}
+
+void nvme_ctrl_free(struct nvme_ctrl *cnt)
+{
+	uint8_t i = 0;
+	if (cnt == NULL)
+		return;
+
+	for (i = 0; i <= nvme_ctrl_npss_get(cnt); ++i)
+		free(cnt->psds[i]);
+	free(cnt->psds);
+
+	free((char *) cnt->dev_path);
+	free((char *) cnt->strs.ver);
+	free((char *) cnt->strs.sn);
+	free((char *) cnt->strs.mn);
+	free((char *) cnt->strs.fr);
+	free((char *) cnt->strs.subnqn);
```

Comment:
`free` takes a `void *`, wouldn't it be more appropriate to cast to `void *` instead of just casting away the constness?